File: packages/agentic/desktop-driver/native/windows/src/host.cpp
#include "host.h"

#include <fcntl.h>
#include <io.h>

#include <algorithm>
#include <iostream>
#include <thread>

#include <winrt/base.h>

#include "build_info.h"

namespace furn {
namespace {

void WriteDiagnostic(const std::string& message) {
  const std::string bounded = message.size() > 1000 ? message.substr(0, 1000) + "..." : message;
  std::cerr << bounded << '\n';
}

}  // namespace

json::Value CreateHello() {
  json::Value hello = json::Value::Object();
  hello.Set("type", json::Value::String(std::string("hello")));
  hello.Set("provider", json::Value::String(std::string("windows")));
  hello.Set("architecture", json::Value::String(std::string("x64")));
  hello.Set("buildId", json::Value::String(std::wstring_view(FurnDesktopDriverBuildId)));
  hello.Set("sourceDigest", json::Value::String(std::wstring_view(FurnDesktopDriverSourceDigest)));
  hello.Set("minimumOs", json::Value::String(std::string("10.0.22000.0")));

  json::Value protocol = json::Value::Object();
  protocol.Set("major", json::Value::Integer(1));
  protocol.Set("minor", json::Value::Integer(0));
  hello.Set("protocol", protocol);

  json::Value features = json::Value::Array();
  for (const char* feature : {"accessibilityClick", "activeElement", "capture", "elementCapture", "find", "hitTest",
                              "keyboard", "launch", "attach", "pointer", "probe", "releaseActions", "setWindowRect",
                              "source", "tree", "wheel"}) {
    features.Append(json::Value::String(std::string(feature)));
  }
  hello.Set("features", features);
  return hello;
}

void Host::WriteResponse(const std::string& id, const CommandResult& result) {
  json::Value response = json::Value::Object();
  response.Set("type", json::Value::String(std::string("response")));
  response.Set("id", json::Value::String(id));
  response.Set("result", result.result);
  if (result.hasBinary) {
    response.Set("binary", result.binaryMetadata);
  }
  writer_->WriteJson(response);
  if (result.hasBinary) {
    writer_->WriteBinary(result.binaryId, result.binaryData);
  }
}

void Host::WriteError(const std::string& id, const std::string& command, const std::string& code,
                      const std::string& message) {
  json::Value error = json::Value::Object();
  error.Set("code", json::Value::String(code));
  error.Set("message", json::Value::String(message));
  json::Value data = json::Value::Object();
  data.Set("command", json::Value::String(command));
  error.Set("data", data);

  json::Value response = json::Value::Object();
  response.Set("type", json::Value::String(std::string("response")));
  response.Set("id", json::Value::String(id));
  response.Set("error", error);
  writer_->WriteJson(response);
}

void Host::WriteCancelled(const std::string& id) {
  json::Value cancelled = json::Value::Object();
  cancelled.Set("type", json::Value::String(std::string("cancelled")));
  cancelled.Set("id", json::Value::String(id));
  writer_->WriteJson(cancelled);
}

void Host::HandleCancel(const std::string& id) {
  {
    const std::lock_guard<std::mutex> guard(mutex_);
    if (activeId_ == id) {
      // The worker owns the acknowledgement: it must stop side effects and
      // release depressed input before the cancellation is reported.
      token_.Cancel();
      return;
    }
    const auto queued = std::find_if(queue_.begin(), queue_.end(),
                                     [&id](const PendingRequest& request) { return request.id == id; });
    if (queued != queue_.end()) {
      queue_.erase(queued);
    }
  }
  // A queued or already-settled request has no running side effects, so the
  // acknowledgement can be written immediately.
  WriteCancelled(id);
}

void Host::WorkerLoop() {
  winrt::init_apartment(winrt::apartment_type::multi_threaded);
  while (true) {
    PendingRequest request;
    {
      std::unique_lock<std::mutex> guard(mutex_);
      signal_.wait(guard, [this] { return stopping_ || !queue_.empty(); });
      if (stopping_ && queue_.empty()) {
        break;
      }
      request = std::move(queue_.front());
      queue_.pop_front();
      token_.Reset();
      activeId_ = request.id;
    }

    bool cancelled = false;
    try {
      const CommandResult result = driver_.Execute(request.command, request.params, token_);
      WriteResponse(request.id, result);
    } catch (const CancelledError&) {
      cancelled = true;
    } catch (const HelperError& error) {
      WriteError(request.id, request.command, error.code(), error.what());
    } catch (const winrt::hresult_error& error) {
      WriteError(request.id, request.command, kErrorInternal, ToUtf8(error.message().c_str()));
    } catch (const std::exception& error) {
      WriteError(request.id, request.command, kErrorInternal, error.what());
    }

    if (cancelled) {
      driver_.ReleaseInput();
      WriteCancelled(request.id);
    }

    {
      const std::lock_guard<std::mutex> guard(mutex_);
      activeId_.clear();
    }
    if (request.command == "dispose") {
      // The session is over: flush the acknowledged response and end the
      // process instead of leaving the reader blocked on a dead stream.
      FlushFileBuffers(GetStdHandle(STD_OUTPUT_HANDLE));
      ExitProcess(0);
    }
  }
  winrt::uninit_apartment();
}

int Host::Run() {
  _setmode(_fileno(stdin), _O_BINARY);
  _setmode(_fileno(stdout), _O_BINARY);
  FrameWriter writer(GetStdHandle(STD_OUTPUT_HANDLE));
  writer_ = &writer;
  FrameReader reader(GetStdHandle(STD_INPUT_HANDLE));

  writer.WriteJson(CreateHello());
  std::thread worker(&Host::WorkerLoop, this);

  int exitCode = 0;
  try {
    Frame frame;
    while (reader.Read(frame)) {
      if (frame.type != kJsonFrameType) {
        WriteDiagnostic("The native helper ignored an unexpected binary frame on stdin.");
        continue;
      }
      const json::Value message =
          json::Value::Parse(std::string_view(reinterpret_cast<const char*>(frame.payload.data()),
                                              frame.payload.size()));
      const std::string type = message.StringField("type");
      const std::string id = message.StringField("id");
      if (type == "cancel") {
        HandleCancel(id);
        continue;
      }
      if (type != "request" || id.empty()) {
        WriteDiagnostic("The native helper received a message that is not a correlated request.");
        continue;
      }
      PendingRequest request;
      request.id = id;
      request.command = message.StringField("command");
      if (const json::Value* params = message.Find("params"); params != nullptr) {
        request.params = *params;
      } else {
        request.params = json::Value::Object();
      }
      {
        const std::lock_guard<std::mutex> guard(mutex_);
        queue_.push_back(std::move(request));
      }
      signal_.notify_one();
    }
  } catch (const HelperError& error) {
    WriteDiagnostic(std::string("furn-desktop-driver-host: ") + error.what());
    exitCode = 1;
  } catch (const std::exception& error) {
    WriteDiagnostic(std::string("furn-desktop-driver-host: ") + error.what());
    exitCode = 1;
  }

  {
    const std::lock_guard<std::mutex> guard(mutex_);
    stopping_ = true;
    token_.Cancel();
  }
  signal_.notify_all();
  worker.join();
  writer_ = nullptr;
  return exitCode;
}

}  // namespace furn

File: packages/agentic/desktop-driver/native/macos/Sources/DesktopDriverHost/Host.swift
import Foundation
import Darwin

func createHello() -> JSONObject {
  [
    "type": "hello",
    "provider": "macos",
    "architecture": "arm64",
    "buildId": FurnBuildInfo.buildId,
    "sourceDigest": FurnBuildInfo.sourceDigest,
    "minimumOs": "14.0",
    "protocol": [
      "major": 1,
      "minor": 1,
    ],
    "features": [
      "accessibilityClick",
      "activeElement",
      "attach",
      "capture",
      "elementCapture",
      "find",
      "focus",
      "hitTest",
      "keyboard",
      "launch",
      "pointer",
      "probe",
      "releaseActions",
      "setWindowRect",
      "source",
      "tree",
      "wheel",
    ],
  ]
}

private struct PendingRequest {
  let id: String
  let command: String
  let params: JSONObject
}

final class Host {
  private let condition = NSCondition()
  private let token = CancellationToken()
  private let driver = Driver()
  private var queue: [PendingRequest] = []
  private var activeID: String?
  private var stopping = false
  private var writer: FrameWriter?
  private let workerDone = DispatchSemaphore(value: 0)

  func run() -> Int32 {
    let writer = FrameWriter(.standardOutput)
    self.writer = writer
    let reader = FrameReader(.standardInput)
    do {
      try writer.writeJSON(createHello())
    } catch {
      writeDiagnostic("furn-desktop-driver-host: writing hello failed: \(error)")
      return 1
    }

    let worker = Thread { [weak self] in
      self?.workerLoop()
    }
    worker.name = "furn-desktop-driver-worker"
    worker.start()

    var exitCode: Int32 = 0
    do {
      while let frame = try reader.read() {
        if frame.type != .json {
          writeDiagnostic("The native helper ignored an unexpected binary frame on stdin.")
          continue
        }
        let value: Any
        do {
          value = try parseJSON(frame.payload)
        } catch {
          writeDiagnostic("The native helper ignored malformed JSON: \(error)")
          continue
        }
        guard let message = value as? JSONObject else {
          writeDiagnostic("The native helper ignored a JSON message that is not an object.")
          continue
        }
        let type = message["type"] as? String ?? ""
        let id = message["id"] as? String ?? ""
        if (type == "cancel" || type == "request")
          && (id.isEmpty || id.utf8.count > Framing.maximumCorrelationIDBytes)
        {
          writeDiagnostic("The native helper ignored a request with an invalid correlation identifier.")
          continue
        }
        if type == "cancel" {
          handleCancel(id)
          continue
        }
        guard type == "request" else {
          writeDiagnostic("The native helper received a message that is not a correlated request.")
          continue
        }
        let command = message["command"] as? String ?? ""
        let params = message["params"] as? JSONObject ?? [:]
        condition.lock()
        queue.append(PendingRequest(id: id, command: command, params: params))
        condition.signal()
        condition.unlock()
      }
    } catch {
      writeDiagnostic("furn-desktop-driver-host: \(error)")
      exitCode = 1
    }

    condition.lock()
    stopping = true
    token.cancel()
    condition.broadcast()
    condition.unlock()
    workerDone.wait()
    self.writer = nil
    return exitCode
  }

  private func handleCancel(_ id: String) {
    condition.lock()
    if activeID == id {
      token.cancel()
      condition.unlock()
      return
    }
    if let index = queue.firstIndex(where: { $0.id == id }) {
      queue.remove(at: index)
    }
    condition.unlock()
    writeCancelled(id)
  }

  private func workerLoop() {
    defer { workerDone.signal() }
    while true {
      condition.lock()
      while queue.isEmpty && !stopping {
        condition.wait()
      }
      if stopping && queue.isEmpty {
        condition.unlock()
        return
      }
      let request = queue.removeFirst()
      activeID = request.id
      token.reset()
      condition.unlock()

      var cancelled = false
      do {
        let result = try driver.execute(command: request.command, params: request.params, token: token)
        try writeResponse(id: request.id, command: request.command, result: result)
      } catch is CancelledError {
        cancelled = true
      } catch let error as HelperError {
        writeError(id: request.id, command: request.command, code: error.code, message: error.message, data: error.data)
      } catch {
        writeError(
          id: request.id,
          command: request.command,
          code: ErrorCode.internalError,
          message: error.localizedDescription
        )
      }

      if cancelled {
        driver.releaseInput()
        writeCancelled(request.id)
      }
      condition.lock()
      activeID = nil
      condition.unlock()

      if request.command == "dispose" && !cancelled {
        try? FileHandle.standardOutput.synchronize()
        Darwin.exit(EXIT_SUCCESS)
      }
    }
  }

  private func writeResponse(id: String, command: String, result: CommandResult) throws {
    if let binaryID = result.binaryID, let binaryData = result.binaryData {
      guard 4 + binaryID.utf8.count + binaryData.count <= Framing.maximumBinaryBytes else {
        writeError(
          id: id,
          command: command,
          code: ErrorCode.captureFailed,
          message: "The captured image exceeds the 64 MiB native binary frame limit."
        )
        return
      }
    }
    var response: JSONObject = [
      "type": "response",
      "id": id,
      "result": result.result,
    ]
    if let metadata = result.binaryMetadata {
      response["binary"] = metadata
    }
    let serialized = try serializeJSON(response)
    if serialized.count > Framing.maximumJSONBytes {
      let treeCommand = command == "find" || command == "source" || command == "tree"
      writeError(
        id: id,
        command: command,
        code: treeCommand ? ErrorCode.treeTooLarge : ErrorCode.internalError,
        message: "The native command response exceeds the 8 MiB JSON frame limit."
      )
      return
    }
    try writer?.writeJSON(response)
    if let binaryID = result.binaryID, let binaryData = result.binaryData {
      try writer?.writeBinary(id: binaryID, data: binaryData)
    }
  }

  private func writeError(
    id: String,
    command: String,
    code: String,
    message: String,
    data: JSONObject? = nil
  ) {
    var errorData = data ?? [:]
    errorData["command"] = bounded(command, bytes: 256)
    var response: JSONObject = [
      "type": "response",
      "id": id,
      "error": [
        "code": code,
        "message": bounded(message, bytes: 4096),
        "data": errorData,
      ],
    ]
    do {
      if try serializeJSON(response).count > Framing.maximumJSONBytes {
        response["error"] = [
          "code": code,
          "message": bounded(message, bytes: 4096),
          "data": [
            "command": bounded(command, bytes: 256),
            "diagnosticsTruncated": true,
          ],
        ]
      }
      try writer?.writeJSON(response)
    } catch {
      writeDiagnostic("Native error write failed: \(error)")
      Darwin.exit(EXIT_FAILURE)
    }
  }

  private func writeCancelled(_ id: String) {
    do {
      try writer?.writeJSON([
        "type": "cancelled",
        "id": id,
      ])
    } catch {
      writeDiagnostic("Native cancellation write failed: \(error)")
      Darwin.exit(EXIT_FAILURE)
    }
  }
}
